### DISCUSSION TITLE
火狐的storage.sync怎么修复？

### DISCUSSION THREAD

**Discussion by jc3213** [jc3213/download_with_aria2#85] [General]
`downloads.download()`方法要求绝对路径，但是我的下载地址是从火狐的`downloads.onCreated`事件里提取出来的，验证也是绝对路径但就是报错

![V5XP$99A9{DWNJC@Q72QO0T](https://github.com/jc3213/download_with_aria2/assets/8744936/b5b4bc90-5da8-4633-b703-8fb0a34d2d19)

这里是[官方文档](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/downloads/download)
测试用扩展 [firefox test.zip](https://github.com/jc3213/download_with_aria2/files/14275959/firefox.test.zip)
`downloads.onCreated`事件也是下载已经开始了，弹出保存文件对话框后这个事件以后才会触发。显得非常的蠢

---------------------

再就是经典的`storage.sync.set`跟`storage.sync.get`无法正常工作了。
这里是[官方文档](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/sync)

![{AJBLM252UEXYAT}B1G@21Q](https://github.com/jc3213/download_with_aria2/assets/8744936/0902d918-3a26-4fec-b6df-6d420fc6ddcb)

**Comment by jc3213** [comment]
@ivysrono @wordpure 谁能去找Moziila的人反馈下，我是真的被整恶心了。

**Reply by ivysrono** [reply]
都是没钱惹的祸……这样分开处理也不错哈。

**Comment by wordpure** [comment]
> `downloads.download()`方法要求绝对路径，但是我的下载地址是从火狐的`downloads.onCreated`事件里提取出来的，验证也是绝对路径但就是报错
> 
> ![V5XP$99A9{DWNJC@Q72QO0T](https://private-user-images.githubusercontent.com/8744936/304641286-b5b4bc90-5da8-4633-b703-8fb0a34d2d19.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc5MjUyNjEsIm5iZiI6MTcwNzkyNDk2MSwicGF0aCI6Ii84NzQ0OTM2LzMwNDY0MTI4Ni1iNWI0YmM5MC01ZGE4LTQ2MzMtYjcwMy04ZmIwYTM0ZDJkMTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTRUMTUzNjAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2ViNmZkY2MxNTQ5OGMxZmZiZTlmODZlYjU4NWI1ZDRkOTU4NTYxMDUxMzkzMzg0OTU0NzE5MmNmM2QxNTgyNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.z_buJ_-OtPjpC43SBsrZjL5WjlkJbPY1ZbDsiuz4a2k)

filename 要传入的应该是一个文件名吧？但看控制台输出的是一个完整的文件路径，换成文件名试试

> 再就是经典的`storage.sync.set`跟`storage.sync.get`无法正常工作了。 这里是[官方文档](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/sync)
> 
> ![{AJBLM252UEXYAT}B1G@21Q](https://private-user-images.githubusercontent.com/8744936/304641771-0902d918-3a26-4fec-b6df-6d420fc6ddcb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc5MjUyNjEsIm5iZiI6MTcwNzkyNDk2MSwicGF0aCI6Ii84NzQ0OTM2LzMwNDY0MTc3MS0wOTAyZDkxOC0zYTI2LTRmZWMtYjZkZi02ZDQyMGZjNmRkY2IucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTRUMTUzNjAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWI0NmE3Y2YzZGIyMmIxYzdlNGU1ZjA3NTliNzYzYzdkZjdjOTBmMzQwYjYxYTdkNzA3N2VlZjQ3MTA5MzNjMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Pdpb7-k7R_iQfspc5LpIaPVOE4FUhbCVbzhouw3Mjok)

`storage.sync.set` 这个，会不会是存入的大小超限了？如果存入失败了，那获取自然也不会成功。

![图片](https://github.com/jc3213/download_with_aria2/assets/68013962/b3c75f96-9a10-4046-bc73-16f24085003b)

> 谁能去找Moziila的人反馈下，我是真的被整恶心了。

可以到 [Mozilla Discourse](https://discourse.mozilla.org/) 反馈看看

**Reply by jc3213** [reply]
`browser.storage.sync.set({test: true})` 也是报错，根本就完全不可用。哪怕最新的Developer版本都不行。就是报错，位置还是undefined，排查都排查不了，脑瘫东西。

`browser.downloads.download({url: "https://github.com/jc3213/download_with_aria2/archive/refs/heads/main.zip", filename: "R:\download_with_aria2-main.zip"})` 一样报错，提示filename必须给绝对地址，我这TM还不是绝对地址？什么真的无语。

**Reply by jc3213** [reply]
主要我的账号因为错误的2FA没了，懒得重新注册，Mozilla当年开始作死的做WebExtension API我就猜到结局了。一样的代码不一样的实现，底层还十分弱智。更别提偶尔遇到的左右不兼容。

本来就被4x版本逼着换Chrome，然后7x版本回来测试，发现看直播直接内存boom。除了部分人还在沉迷的userChrome外，几乎就是个完全劣化版本的Chrome。或许纯净状态下速度更快，但是加上扩展就是或多或少的问题。部分API又保持了原本的XUL框架下的功能，但是抄袭Chromium部分的真的一言难尽。

**Comment by wordpure** [comment]
> `browser.downloads.download({url: "https://github.com/jc3213/download_with_aria2/archive/refs/heads/main.zip", filename: "R:\download_with_aria2-main.zip"})` 一样报错，提示filename必须给绝对地址

`filename must not be an absolute patn` 的意思是 `文件名不能是绝对路径` 😂

> `browser.storage.sync.set({test: true})` 也是报错，根本就完全不可用。哪怕最新的Developer版本都不行。就是报错，位置还是undefined，排查都排查不了

不是很懂这个，不过我在 4.7.0.2442 的 `pages/options/options.html` 测试是正常的

![图片](https://github.com/jc3213/download_with_aria2/assets/68013962/f4e993f8-0aaa-4716-98f6-8670b9b685c9)

**Reply by wordpure** [reply]
顺带一提，如果确实很反感 Firefox 的话，可以在 README 里说明一下扩展只支持 Chromium，然后放弃对 Firefox 的兼容

**Reply by jc3213** [reply]
为什么你就可以正常工作，我这边就直接提示错误？难道是因为我没登录Firefox账号？还是说因为我使用的版本没有签名？我就记得当初我给火狐也写的sync然后我自己这边测试失败了以后两边都改成了local。

`文件名不能是绝对路径`  我的，眼睛花了。不得不说火狐暗色模式下提示词五颜六色的还真不容易看清楚。加上昨天实测新功能逻辑上也有问题，最后还是返回最初的功能了。

**Reply by jc3213** [reply]
我自己下载 4.7.0.2442 然后封装了一个火狐版本，然后各个子页面就不正常了。我是真的无语了。就是用着有问题。如果Developer版本的浏览器跟未签名版本的扩展无法正常提供功能，那我不知道说什么好了

**Reply by jc3213** [reply]
火狐使用sendResponse(即我的response)就会出错，用return就正常工作，实在太秀了。有一说一。https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#sendresponse

---------------------

测试了下貌似是跟我标记了async有关，但是理论上不管这是否是async function，response是一个callback应该正常工作啊？删除了async就正常了，似乎火狐在执行async response的时候必须`return true`或者`return new Promise`。看起来这个问题是Chromium这边的，不过问题解决了就行了，懒得纠结。

现在`storage.sync`那边还是解决不了，即使从AMO安装有签名的版本还是报错，简直有毒

**Reply by jc3213** [reply]
不得不说Chromium MV3的`Service Worker`非常不适合这个扩展，但是现在好歹有hacks可以拥有persistent service worker，至少不会太过扭曲了。否则会陷入多次`chrome.storage.sync.get`，以及属性的初始化。

Firefox MV3则直接禁止非SSL连接，也就是只要升级了，大多数本地使用的人都不得不用OpenSSL去生成一个pem。

两家的Manifest V3都挺蛋疼的。V2的话问题还行。就是chrome API不是Promise，操作起来还挺麻烦

**Comment by jc3213** [comment]
@wordpure 为什么你那边`storage.sync`就正常，我这边就出错……真的无解，全新档案也是错，头皮发麻
![无语](https://github.com/jc3213/download_with_aria2/assets/8744936/34a34438-0a29-4b5a-9abd-f07df41aa2be)

**Reply by wordpure** [reply]
> 为什么你就可以正常工作，我这边就直接提示错误？难道是因为我没登录Firefox账号？

`storage.sync` 应该要登录 Firefox 账号之后才会生效，你登录下试试

![图片](https://github.com/jc3213/download_with_aria2/assets/68013962/4a1034c7-f397-4f37-8ae9-5f2dc748972f)

**Reply by jc3213** [reply]
浏览器本身的储存区域不应该强制登陆才能访问吧？Chromium系也没有强制登陆才能访问啊？这也难怪我自己跟我朋友测试之前发布的storage.sync版本都出错，但是大多数人都没反馈有问题。

关键火狐这个错误提示也太过离谱了，错误位置undefined，错误内容：发生错误。真是想解决都解决不了。唉

**Reply by jc3213** [reply]
关键文档也少，讨论也少。除了可能的是否登陆火狐的同步账号外几乎没其他的可疑点。头疼

**Reply by wordpure** [reply]
> 浏览器本身的储存区域不应该强制登陆才能访问吧？Chromium系也没有强制登陆才能访问啊？

`storage.local` 才是本地存储，`storage.sync` 是浏览器云端同步，应该都要关联账号才能使用的。可以试试在 Chromium 上把对应的谷歌、微软账号退掉，看 `storage.sync` 还能不能正常使用。

![图片](https://github.com/jc3213/download_with_aria2/assets/68013962/d2a7892c-4abd-4e36-b67f-0fe2825e629d)

**Reply by jc3213** [reply]
不适用`-profile "dir"`就正常了。然而完全不符合我的使用习惯，人麻了。

**Reply by wordpure** [reply]
不用 `-profile "dir"`，就可以在不登录 Firefox 账号的情况下启用 `storage.sync` ？

还是说你只在默认的 profile 里登录了账号，其他 profile 都没有登录，所以才用不了的？

**Reply by jc3213** [reply]
不用`-profile "dir"`就可以了，不用登陆也能使用。但是用了`-profile "dir"'`，虽然能转移用户档案的位置，然后直接boom了。使用`storage.sync`就会提示发生预期外的错误，位置undefined。关键这个完全没任何文档提到，stackoverflow也没有任何说法，问Bing AI甚至翻出chrome那边的对策方案（关键一眼错）。

**Reply by wordpure** [reply]
那就怪了，不用登录的话，它是怎么跨设备同步的？

**Reply by jc3213** [reply]
你甚至可以自己用`-profile "dir"`自己新建一个全空的档案，然后装上最新版本，你就发现储存相关全部boom了。但是直接运行firefox.exe就正常。因为即使是`sync`空间，本质是先储存在本地，然后再跟服务器同步，无法写入这个行为就异常奇怪

**Comment by jc3213** [comment]
你猜我怎么解决的这个问题的？这个错误之所以会发生是因为不能使用`-profile "profile dir"`这个参数。不知道Firefox Portable Launcher会否报错。就真的很迷糊。我专门用多个快捷方式配合`-profile`设置多个不同的用户档案用于测试一些本地玩具，反而还弄巧成拙了。离谱！ @wordpure 

![20240215195916](https://github.com/jc3213/download_with_aria2/assets/8744936/e76673dd-a179-4983-82a0-19700ed3d3c0)
